### PR TITLE
fix(coding-agent): paste clipboard images on macOS Cmd+V

### LINKED ISSUE #2186: Support native macOS Cmd+V for clipboard images
## Problem

On macOS, pasting a screenshot or other clipboard image into Atomic works with `Ctrl+V`, but the native macOS paste command, `Cmd+V`, appears to do nothing. This is surprising because Mac users expect `Cmd+V` to paste both text and images.

## Why this happens

Atomic binds `app.clipboard.pasteImage` to `Ctrl+V` on macOS/Linux. That key calls the existing clipboard-image reader directly.

`Cmd+V` takes a different path: the terminal consumes the shortcut and sends Atomic a bracketed-paste sequence. Text arrives inside that sequence and works normally. When the clipboard holds only an image, the sequence has an empty text payload (`\x1b[200~\x1b[201~`). `@earendil-works/pi-tui`'s base `Editor` ignores an empty paste, so Atomic never calls its image reader.

This behavior also matches the upstream report in earendil-works/pi#4332.

## Proposed minimal fix

In `packages/coding-agent/src/modes/interactive/components/custom-editor.ts`, on macOS only (`process.platform === "darwin"`), detect an empty bracketed-paste event before passing input to the base editor and call the existing `onPasteImage` handler—the same path used by `Ctrl+V`.

Keep this behavior strictly isolated to macOS. Do not change Linux or Windows paste handling or default keybindings. Adding a `Cmd+V`/`super+v` keybinding would not solve the common case because the terminal owns `Cmd+V` and sends a paste event rather than a normal key event.

## Acceptance criteria

- On macOS, `Cmd+V` with a screenshot or image on the clipboard attaches/inserts the image exactly as `Ctrl+V` does today.
- `Cmd+V` with clipboard text keeps using normal bracketed-paste behavior.
- The existing `Ctrl+V` image-paste shortcut remains unchanged.
- On macOS, an empty bracketed-paste event invokes the image-paste handler once and does not insert control bytes.
- On Linux and Windows, empty bracketed-paste handling and default keybindings remain unchanged.
- Add focused tests for macOS empty and non-empty bracketed-paste input, plus a non-macOS isolation case, in `CustomEditor`.
- Update `packages/coding-agent/docs/keybindings.md` to state that native macOS `Cmd+V` can paste clipboard images, while `Ctrl+V` remains the explicit `app.clipboard.pasteImage` binding.

## Likely scope

A small macOS-only `CustomEditor.handleInput()` guard, focused unit tests (including non-macOS isolation), and one docs update. No new clipboard reader or native module should be needed.

### PR DESCRIPTION
**Pull request by Shreyasd10**
## Summary

On macOS, `Ctrl+V` could paste a screenshot into Atomic, but native `Cmd+V` often did nothing. That is the shortcut Mac users actually hit, so it felt broken even though the explicit image-paste binding worked fine.

[#2186](https://github.com/bastani-inc/atomic/issues/2186) has the full story. Short version: the terminal owns `Cmd+V`. For an image-only clipboard it usually does not send ordinary text paste. Some terminals send an empty bracketed-paste sequence (`ESC[200~` right next to `ESC[201~`). Kitty-protocol terminals like Ghostty send a `super+v` key event instead. pi-tui's base editor ignores empty paste, and Atomic never treated either signal as image paste, so `onPasteImage` never ran.

This PR does not touch the clipboard reader or the `app.clipboard.pasteImage` binding. It only teaches chat `CustomEditor` to recognize those two macOS deliveries and call the same `onPasteImage` path `Ctrl+V` already uses. Fixes #2186.

## Scope vs #2186

The issue asked for a small macOS `CustomEditor.handleInput()` guard for empty bracketed paste, plus tests and docs. No new clipboard reader, no native module.

That covers terminals that still send Atomic *something* for image-only `Cmd+V` (empty paste or Kitty `super+v`). It does not cover Apple Terminal.app when image-only `Cmd+V` produces no PTY event at all. Fixing that would need an OS-level key intercept or a non-PTY input path, which is outside #2186.

The acceptance text says "on macOS, `Cmd+V`…" without naming a terminal. I read that as: wherever the host forwards the gesture, Atomic should paste the image. Terminal.app's silent drop is a host limit; we document it instead of pretending this PR fixes it.

## What changed

In `packages/coding-agent/src/modes/interactive/components/custom-editor.ts`, `handleInput` still checks extension shortcuts first. After that, a paste-image signal is either:

- the configured `app.clipboard.pasteImage` keybinding (`Ctrl+V` on macOS/Linux, `Alt+V` on Windows), or
- on darwin only, an exact empty bracketed-paste event, or Kitty-protocol `super+v` via `matchesKey`

Both call `onPasteImage?.()` and return without handing the bytes to the base editor, so empty paste does not leak control sequences into the draft.

Exact equality with `\x1b[200~\x1b[201~` is intentional. Partial or embedded sequences stay ordinary input. A looser matcher would risk swallowing adjacent text.

Ghostty can emit Kitty repeat events while the key is held. Without a guard, one held `Cmd+V` stacked duplicate images. The paste-image branch now skips `isKeyRepeat(data)` for every paste-image signal (including held `Ctrl+V`), then still consumes the event so repeats do not fall through as editor input.

Docs and changelog:

- `docs/keybindings.md` and `docs/quickstart.md` cover native macOS `Cmd+V` for images, that it is not a bindable Atomic shortcut, the Apple Terminal caveat, and that mixed text+image clipboards can disagree depending on how the terminal delivered the gesture
- `[Unreleased]` Fixed notes the user-facing fix without bracketed-paste implementation detail

## Before / after

Before:

- Image-only `Cmd+V` in VS Code's terminal: empty paste, draft unchanged, no attachment
- Image-only `Cmd+V` in Ghostty: `super+v` ignored for image paste
- Held paste key could insert the same image repeatedly
- Docs only mentioned `Ctrl+V` / `Alt+V`

After:

- macOS empty bracketed paste calls `onPasteImage` once and leaves a seeded draft unchanged
- macOS Kitty `super+v` takes the same path
- Linux/Windows empty paste and Linux `super+v` still do not call the image handler
- Repeat events are consumed without a second paste
- Extension shortcuts still win when they claim the same input
- Terminal.app image-only `Cmd+V` still does nothing when the host sends no event; `Ctrl+V` remains the workaround there

## Apple Terminal.app

Manual check: image-only `Cmd+V` in Terminal.app often never reaches Atomic. No empty paste, no `super+v`, no key event on the PTY. This PR cannot route what it never receives.

Workaround there: `Ctrl+V` (explicit `app.clipboard.pasteImage`), or use Ghostty / iTerm / Kitty / VS Code's terminal, where the host does forward a signal.

Making Terminal.app `Cmd+V` work for images would be a separate follow-up (native event tap or non-PTY UI), not an extension of this `CustomEditor` guard.

## Tests

`packages/coding-agent/test/custom-editor-macos-paste.test.ts` drives a real `CustomEditor` with a platform stub and a call counter (no OS clipboard). It covers:

- non-empty macOS bracketed paste inserts text and does not call `onPasteImage`
- empty bracketed paste on linux / win32 does not call `onPasteImage`
- explicit `pasteImage` keybinding still routes once, including when the handler is missing
- partial / embedded empty-paste sequences are not treated as image paste
- exact empty paste on darwin calls the handler once without changing the draft
- extension shortcuts beat empty paste and `super+v` on darwin
- Kitty `super+v` on darwin routes once; repeat events do not
- Kitty `super+v` on linux does not call the handler
- Kitty repeat on the explicit `ctrl+v` binding does not re-fire paste

## Verification

| Command | Outcome |
| --- | --- |
| `npm run test --workspace=@bastani/atomic -- test/custom-editor-macos-paste.test.ts` | pass |
| Manual: Ghostty image-only `Cmd+V` | one image via Kitty `super+v` |
| Manual: VS Code integrated terminal image-only `Cmd+V` | one image via empty bracketed paste |
| Manual: `Ctrl+V` image paste | unchanged |
| Manual: Apple Terminal.app image-only `Cmd+V` | no PTY event (out of scope); `Ctrl+V` still works |

## Notes

- Scope is chat `CustomEditor` only. No new clipboard reader, native module, or pi-tui `onEmptyPaste` hook.
- A user-facing `cmd+v` / `super+v` keybinding would not fix the common case: the terminal often never delivers a normal keybinding event for native paste.
- Mixed clipboard: empty-paste terminals may still insert text on `Cmd+V` while `Ctrl+V` prefers the image. Ghostty-style `super+v` goes through the image path, so it can prefer the image the same way `Ctrl+V` does. Docs mention that instead of promising one universal split.

<h3>Greptile Summary</h3>

This update adds macOS image-paste routing for empty bracketed-paste and Kitty `super+v` input, prevents repeated terminal key events from invoking image paste again, and documents terminal-specific Cmd+V behavior. The focused editor suite passes all 14 routing, shortcut-precedence, and repeat-event checks with the runtime `TuiMainScreen` fixture.

<h3>Confidence Score: 5/5</h3>

No blocking failure remains.

The focused macOS paste suite passed all 14 checks, including native signal routing, extension shortcut precedence, repeat-event suppression, and fixture construction with the installed terminal UI runtime.

<details><summary><h3><a href="https://www.greptile.com/trex">T-Rex</a> T-Rex Logs</h3></summary>

**What T-Rex did**
- T-Rex completed the requested verification but did not upload local artifact references.
- T-Rex ran the focused fixture against the installed runtime, which produced 14 failures due to TUI not being a constructor, and then ran the updated focused suite with TuiMainScreen, which passed all 14 tests.
- T-Rex again ran the requested verification but did not upload local artifact references.
- T-Rex executed two exact Vitest commands for the macOS paste tests: the paste-before tests failed with 14 failures caused by TypeError: TUI is not a constructor, and the custom-editor tests passed with 14 tests across 1 file; no new executable validation source was authored.

<a href="https://app.greptile.com/trex/runs/18437883/artifacts"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/ViewAllArtifactsDark.svg?v=4"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/ViewAllArtifacts.svg?v=4">View all artifacts</picture></a>

<sub><a href="https://www.greptile.com/trex">T-Rex</a> Ran code and verified through T-Rex</sub>
</details>

<sub>Reviews (7): Last reviewed commit: ["fix(coding-agent): finish macOS image pa..."](https://github.com/bastani-inc/atomic/commit/376246edc2019146666c45c19704ed3e607982e7) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=50714409)</sub>

### CODE REVIEW

File: packages/coding-agent/test/custom-editor-macos-paste.test.ts [hunk 1]
```diff
@@ -0,0 +1,294 @@
+import { type Terminal, TUI } from "@earendil-works/pi-tui";
+import { beforeAll, describe, expect, it } from "vitest";
+import { KeybindingsManager } from "../src/core/keybindings.ts";
+import { CustomEditor } from "../src/modes/interactive/components/custom-editor.ts";
+import { getEditorTheme, initTheme } from "../src/modes/interactive/theme/theme.ts";
```
[the remark this answers is not in the document]
Comment (Shreyasd10) [reply packages/coding-agent/test/custom-editor-macos-paste.test.ts:5]: Not valid for this package. `packages/coding-agent/test` already uses `.ts` local imports (hundreds of files, including `custom-editor-prompt.test.ts` / `custom-editor-placeholder.test.ts`, which this file mirrors). The AGENTS.md `.js` ESM note is for Bun-resolved shipped source elsewhere; switching these three lines to `.js` would be the inconsistency. Leaving as-is.

**Comment by Shreyasd10** [comment]
Hey, circling back on this one. It's been sitting for a few days with no maintainer eyes on it.

Conflicts with main are fixed, CI is green, and Greptile already cleared it. Scope is still just the #2186 CustomEditor path: empty bracketed paste and Kitty `super+v` on macOS route into `onPasteImage`. Terminal.app still drops image-only Cmd+V with no PTY event; that's documented as out of scope.

Would appreciate a look when you get a minute.

**Comment by flora131** [comment]
Short recording of a hands-on manual run of this PR's locally built Atomic TUI inside tmux after rebasing onto current `main` (`bffe6d8771c6`). The clip shows local head `2eb4e2cb2c60`, the native macOS Cmd+V image-paste interaction, and the observed result.

https://github.com/user-attachments/assets/4748b410-c972-4def-8373-8381da91cfd3

**Comment by flora131** [comment]
Thanks @Shreyasd10 for the fix and the coverage — we tested the feature and it works.

We also rebased the PR onto the latest `main` and updated the test fixture to fix the failures exposed by the latest pi-tui changes. The focused suite now passes all 14 tests, and the repository checks pass.

Inside tmux on macOS, `Ctrl+V` is the reliable image-paste path. Native `Cmd+V` depends on terminal forwarding: VS Code's terminal may forward the empty bracketed-paste route through tmux, while Ghostty+tmux may not forward its Kitty `super+v` route. This is a terminal forwarding limitation, not an Atomic defect.

The real manual TUI recording is here: https://github.com/bastani-inc/atomic/pull/2211#issuecomment-5260497988. You retain credit for this PR, and we plan to merge it with the next set of changes for the release.
